File: softmax_test/softmax.py
import numpy as np

def read_data(file_path):
    with open(file_path, 'r') as file:
        data = file.readlines()
    return np.array([float(line.strip()) for line in data])

def softmax(x):
    max = 0
    pre_max = max
    ans = []
    for i in range(0,(1024*768),768):
        # if(i<2048):
        #     print(i)
        #     print(x[i:i+16])
        a = np.exp(x[i:i+768] - np.max(x[i:i+768]))/np.sum(np.exp(x[i:i+768] - np.max(x[i:i+768])))
        # a = np.exp(x[i:i+768])/np.sum(np.exp(x[i:i+768]))
        for n in a:
            ans.append(n)
    return ans

def verify_softmax(calculated, expected, tolerance=1e-6):
    return np.allclose(calculated, expected, atol=tolerance)

data = read_data('random_floats.txt')
# data = read_data('random_floats_16bit_fixed_point.txt')
calculated_softmax = softmax(data)
# np.savetxt('expected_softmax.txt', calculated_softmax)
np.savetxt('expected_softmax_safe.txt', calculated_softmax)
# np.savetxt('expected_softmax_16bit_fixed_point.txt', calculated_softmax)
# expected_softmax = read_data('expected_softmax.txt')
# is_correct = verify_softmax(calculated_softmax, expected_softmax)

# if(is_correct == False):
#     for i in range(len(calculated_softmax)):
#         if(calculated_softmax[i] != expected_softmax[i]):
#             print('i = {}, calculated_softmax = {}, expected_softmax = {}'.format(i, calculated_softmax[i], expected_softmax[i]))

# print("Calculated Softmax:", calculated_softmax)
# print("Expected Softmax:", expected_softmax)
# print("Softmax calculation is correct:", is_correct)

File: softmax.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>
#include <algorithm>

std::vector<float> read_data(const std::string& file_path) {
    std::ifstream file(file_path);
    std::vector<float> data;
    float value;
    
    while (file >> value) {
        data.push_back(value);
    }

    return data;
}

std::vector<float> softmax(const std::vector<float>& x) {
    std::vector<float> ans;
    int size = 1024 * 768;
    int chunk_size = 768;

    for (int i = 0; i < size; i += chunk_size) {
        // float max_val = *std::max_element(x.begin() + i, x.begin() + i + chunk_size);
        float sum_exp = 0.0, max_val = x[i], pre_max = x[i];
        for (int j = 0; j < chunk_size; ++j) {
            pre_max = max_val;
            max_val = std::max(max_val, x[i + j]);
            sum_exp = sum_exp * std::exp(pre_max - max_val) + std::exp(x[i + j] - max_val);
            if(i == 0){
                printf("pre_max: %f max_val: %f \n",pre_max, max_val);
            }
        }
        if(i == 0){
            printf("%f\n",sum_exp);
        }
        for (int j = 0; j < chunk_size; ++j) {
            
            ans.push_back(std::exp(x[i + j] - max_val) / sum_exp);
        }
    }
    return ans;
}

void save_data(const std::string& file_path, const std::vector<float>& data) {
    std::ofstream file(file_path);
    for (float val : data) {
        file << val << "\n";
    }
}

int main() {
    std::string input_file = "random_floats.txt";
    std::string output_file = "expected_softmax_online_float.txt";

    std::vector<float> data = read_data(input_file);
    std::vector<float> calculated_softmax = softmax(data);
    save_data(output_file, calculated_softmax);

    return 0;
}
